pytorch_lightning/metrics/metric.py

```diff
@@ -0,0 +1,121 @@
+import numbers
+from abc import ABC, abstractmethod
+from collections import Mapping, Sequence
+from functools import partial
+from typing import Union, Any, Optional
+
+import torch
+import torch.distributed
+from torch.utils.data._utils.collate import np_str_obj_array_pattern
+
+__all__ = ['BaseMetric']
+
+
+class BaseMetric(torch.nn.Module, ABC):
+    def __init__(self, name: str,
+                 reduce_group: Optional[Any] = torch.distributed.group.WORLD,
+                 reduce_op: Optional[Any] = torch.distributed.ReduceOp.SUM):
+        """
+        Abstract Base Class for metric implementation.
+
+        Automatically handles the computation
+        Args:
+            name: the metric's name
+            reduce_group: the process group for DDP reduces (only needed for DDP training).
+                Defaults to all processes (world)
+            reduce_op: the operation to perform during reduction within DDP (only needed for DDP training).
+                Defaults to sum.
+        """
+        super().__init__()
+        self.name = name
+        self.reduce_op = reduce_op
+        self.reduce_group = reduce_group
+
+    @abstractmethod
+    def forward(self, *args, **kwargs) -> torch.Tensor:
+        """
+        Implements the actual metric computation.
+
+        Returns:
+            metric value
+
+        """
+        raise NotImplementedError
+
+    def __call__(self, *args, **kwargs) -> torch.Tensor:
+        return sync_collections(super().__call__(*args, **kwargs),
+                                group=self.reduce_group,
+                                reduce_op=self.reduce_op)
+
+
+def sync_ddp(result: Union[torch.Tensor, numbers.Number],
```

Comment:
this is public or private (do we want to expose it to users)?

Reply:
A mixture of both, usually I'd say, this should be private (hence it's not included in `__all__`).
More experienced users might however want to call it directly, which is why I didn't ad a `_` prefix.
But we can also make this private

Reply:
I may add `_` so it clear that is not meant for general usage but you still can if you want...